ChartController.js

```diff
@@ -0,0 +1,148 @@
+*/
+
+/**
+ * Describes the format of the data
+ * @typedef {Object} Format
+ * @property {Prefix} [prefix = ({last}) => last.monitors.os] -
+ * Shortens the number of properties needed to get to a value.
+ * @property {Function} [y] - Required for time/line charts describes the y.
+ * @property {Function} [x = r => r.createdDate] -
+ * Optional function that gets the x value for line charts.
+ * @property {Function} [last] -
+ * Pie Charts use this to calculate the last value.
+ * @property {Function} max - Returns the max value.
+ */
+
+class ChartController {
+  /**
+   * @param {Object} options - Options for the chart.
+   * @param {ChartType} [options.type = 'pie'] - Chart type.
+   * @param {Format} options.format - The format for the chart.
+   * @param {number} [options.poll = 2000] - How often StatsService will poll.
+   */
+  constructor({type = 'pie', format, poll = 2000}) {
+    const chartType = ChartType[type];
+    if(!chartType) {
+      throw new Error(`${type} Charts are unsupported`);
+    }
+    this.type = type;
+    this.format = format;
+    this.format.units = ChartUnit;
+    this._chart = {};
+    this._statsService = new StatsService({poll});
+    this.id = `${type}-${Date.now()}`;
+    this.updater.bind(this);
+    this._statsService.subscribe(this);
+  }
+  /**
+   * @returns {Object} An updated chart.
+   */
+  get chart() {
+    return this._chart;
+  }
+  /**
+   * @returns {ChartUnit} Can be used to format data.
+   */
+  get units() {
+    return ChartUnit;
+  }
+  /**
+   * @returns {boolean} Is the chart loading?
+   */
+  get loading() {
+    return this._statsService.loading;
+  }
+  /**
+   * @returns {boolean} Is the chart updating?
+   */
+  get updating() {
+    return this._statsService.updating;
+  }
+  /**
+   * Constructs the chart using the ChartType.
+   *
+   * @param {Object} data - Data from the StatsService update.
+   * @param {Array} data.latest - The data from the latest update only.
+   * @param {Array<Object>} data.all - All of the data received so far.
+   *
+   * @returns {Object} Chart with new data.
+   */
+  set chart(data) {
+    const {prefix = d => d} = this.format;
+    const root = prefix(data) || data;
+    this._chart.max = this.format.max(root);
+    switch(this.type) {
+      case ChartType.pie: {
+        this._chart.last = this.format.last(root);
+        return this._chart;
+      }
+      case ChartType.line:
+      case ChartType.time:
+      case ChartType.realtime: {
+        const {x = () => data.latest.createdDate} = this.format;
+        this._chart.series = [{
+          x: x(root),
+          y: this.format.y(root)
+        }];
+        return this._chart;
+      }
+      default: {
+        console.warn(`unable to format ${this.type} charts`);
+        return this._chart;
+      }
+    }
+  }
+  /**
+   * Called by StatsService on each update.
+   * This function is async because we do not want it blocking.
```

Comment:
Don't think this comment is necessary. Javascript is single threaded, therefore you almost never want to block the main loop.

Reply:
I actually think this comment is necessary because by being async is goes on the async que which means that the StatsService is able to issue all updates to a que instead of awaiting them.